src/basic/fd-util.c

```diff
@@ -896,6 +896,30 @@ int fd_reopen(int fd, int flags) {
         return new_fd;
 }
 
+int fd_reopen_replace(int *fd, int flags) {
+        assert(fd);
+        assert(*fd >= 0);
+
+        /* Reopens the specified fd with new flags, closing the original fd if the reopen was successful.
+         *
+         * Returns an error if the close() on the original file descriptor fails.  In that case, the original
+         * file descriptor is already gone and *fd already refers to the new descriptor.
+         *
+         * In all cases, before and after the call, *fd will refer to an open file descriptor.
+         *
+         * If the reopen and the close succeeded then 0 is returned, else an error.
+         */
+        int new_fd = fd_reopen(*fd, flags);
+        if (new_fd < 0) {
+                return new_fd;
+        }
+
+        int old_fd = *fd;
+        *fd = new_fd;
+
+        return close_nointr(old_fd);
```

Comment:
```c
close_and_replace(*fd, new_fd);
```

Reply:
close_and_replace() uses safe_close() which uses close_nointr(), hence it's the same pretty much. Please use close_and_replace() hence.

(We generally don't want that any code in our codebase calls plain close() actually, they all should call safe_close() typically, or one of the wrappers of it)